File: haskell/src/Sol/P0026.hs
module Sol.P0026 (compute, solve) where

import Mylib.Factor (divisors, primeFactorization)
import Mylib.Math (powerModExn)
import Mylib.Util (headExn)

pp :: Int -> Int
pp = iterDiv 5 . iterDiv 2
  where
    iterDiv :: Int -> Int -> Int
    iterDiv divisor x
        | x `mod` divisor == 0 = iterDiv divisor (x `div` divisor)
        | otherwise = x

-- This implementation isn't a correct Carmichael function because
-- it assumes that the argument is odd (not a multiple of 2).
carmichael :: Int -> Int
carmichael n =
    foldl1 lcm
        . map (\(x0, x1) -> (x0 - 1) * (x0 ^ (x1 - 1)))
        $ primeFactorization n

findRepetendLength :: Int -> Int
findRepetendLength n
    | d == 1 =
        0
    | otherwise =
        headExn
            . filter ((== 1) . (flip . powerModExn) 10 d)
            . divisors
            $ carmichael d
  where
    d = pp n

compute :: Int -> String
compute limit =
    show $ aux (reverse [(div limit 2) .. (limit - 1)]) 0 0
  where
    aux :: [Int] -> Int -> Int -> Int
    aux [] _ answer = answer
    aux (x : xs) max_length answer
        | x <= max_length =
            answer
        | otherwise =
            let tmp = findRepetendLength x
             in if tmp > max_length
                    then
                        aux xs tmp (pp x)
                    else
                        aux xs max_length answer

solve :: String
solve = compute 1_000

File: go/internal/sol/p0026/p0026.go
package p0026

import (
	"maps"
	"slices"
	"strconv"

	"pe-solver/internal/mylib"
)

// This function is not the correct Carmichael function because
// the function assumes that the argument is not a multiple of 2.
func carmichael(n int) int {
	var result = 1
	for b, e := range maps.All(mylib.Frequencies(mylib.PrimeFactorization(n))) {
		result = mylib.Lcm(result, (b-1)*mylib.Pow(b, (e-1)))
	}

	return result
}

func pp(n int) int {
	for n%2 == 0 {
		n /= 2
	}
	for n%5 == 0 {
		n /= 5
	}

	return n
}

func findRepetendLength(d int) int {
	if d = pp(d); d == 1 {
		return 0
	}

	for k := range slices.Values(mylib.Divisors(carmichael(d))) {
		if mylib.PowerMod(10, k, d) == 1 {
			return k
		}
	}

	panic("unreachable")
}

func compute(limit int) string {
	var maxLength int
	var result int

	for i := limit - 1; i >= limit/2; i-- {
		if i <= maxLength {
			break
		}
		if repLength := findRepetendLength(i); repLength > maxLength {
			result = i
			maxLength = repLength
		}
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute(1_000)
}
